pkg/controlplane/controller/crdregistration/crdregistration_controller.go

```diff
@@ -40,6 +40,8 @@ import (
 type AutoAPIServiceRegistration interface {
 	// AddAPIServiceToSync adds an API service to auto-register.
 	AddAPIServiceToSync(in *v1.APIService)
+	// AddAPIServiceToSyncSynchronously adds an API service to auto-register synchronously without using queue.
+	AddAPIServiceToSyncSynchronously(in *v1.APIService)
```

Comment:
that's an interesting idea

Reply:
Does it work out, i.e. get you get the wiring through without the race?

Reply:
I'm still trying to find a way how to test the idea(currently could not find) and PR is still WIP but my theory is such that;
crdregistration_controller enqueue apiservices in here https://github.com/kubernetes/kubernetes/blob/c964fe6001376c8ad209656175bd3c4087ce5d71/pkg/controlplane/controller/crdregistration/crdregistration_controller.go#L210

However, autoregistration_controller does not start consuming queue until here unleashes https://github.com/kubernetes/kubernetes/blob/c964fe6001376c8ad209656175bd3c4087ce5d71/cmd/kube-apiserver/app/aggregator.go#L150

Meanwhile, apiserviceregistration controller starts consuming https://github.com/kubernetes/kubernetes/blob/c964fe6001376c8ad209656175bd3c4087ce5d71/staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go#L264
and aggregating OpenAPI specs for API services in here https://github.com/kubernetes/kubernetes/blob/c964fe6001376c8ad209656175bd3c4087ce5d71/staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go#L395

However, due to the race between autoregistration and apiserviceregistration controllers, some OpenAPI specs are being missed in first run.

Reply:
@sttts, current change(crdregistration calls autoregister synchronously) does not work in OpenShift. As you worried about deadlock(Apiserver and SDN?), something blocks bootstrap completes installation.